components/sup/src/manager/mod.rs

```diff
@@ -687,6 +713,17 @@ impl Manager {
         if self.http_disable {
             info!("http-gateway disabled");
         } else {
+            // First let's check and see if we're going to use TLS. If so, we'll generate the
+            // appropriate config here, where it's easy to propagate errors, vs in a separate
+            // thread, where that process is more cumbersome.
+            let tls_server_config = match self.state.cfg.tls_files {
+                Some(ref files) => match tls_config(files.clone()) {
```

Comment:
That fails with the following error:

![](https://dl.dropboxusercontent.com/s%2F5e79a1pqecwxle7%2F1.%252000.1%2520fish%25202018-11-30%252013-59-28.png)

I'll work on simplifying this particular bit though.

Reply:
I think the need to return early if `tls_config()` returns an `Err` is what makes simplifying this tricky.  Essentially you want to go from `Option` -> `Result` -> `Option` but return early if there's an error.  I've simplified this a little bit, but if there's a way to go farther with it, let me know.  I don't see it right now.

Reply:
Oh, I see... you can't use `?` inside the lambda function inside the `map` call, because _it_ doesn't return a `Result`. D'oh!

There's probably some way to do it, but my brain is getting mushy this late in the day, and what you've got certainly works.

Reply:
I couldn't resist playing with this a bit. I got it to work, but I'm not sure it's an improvement:
```rust
            let tls_server_config = self
                .state
                .cfg
                .tls_files
                .as_ref()
                .map(|(key, cert)| tls_config(key, cert))
                .map_or(Ok(None), |v| v.map(Some))?;
```

After the `map`, you have a `Option<Result<ServerConfig>>`, but we need to [`transpose`](http://doc.rust-lang.org/1.30.1/std/result/enum.Result.html#method.transpose) it, and because that's not stable yet, we have to do it ourselves to get a `Result<Option<ServerConfig>>`, which we can apply `?` to to propagate the error or unwrap to a `Option<ServerConfig>`.

It's not necessary to have the `ref`s in `(ref key, ref cert)`, but you do need `as_ref` to convert the `Option<(PathBuf, PathBuf)>` to `Option<&(PathBuf, PathBuf)>`, or else you run into this:
```
error[E0507]: cannot move out of borrowed content                                                                                                                                                           
   --> components/sup/src/manager/mod.rs:732:37                                                                                                                                                             
    |                                                                                                                                                                                                       
732 |               let tls_server_config = self                                                                                                                                                            
    |  _____________________________________^                                                                                                                                                               
733 | |                 .state                                                                                                                                                                              
    | |______________________^ cannot move out of borrowed content                                                                                                                                          
                                                                                                                                                                                                            
error: aborting due to previous error                                                                                                                                                                       
                                                                                                                                                                                                            
For more information about this error, try `rustc --explain E0507`.                                                                                                                                         
```
Because the function takes ownership of `self`.